### PR TITLE
overlord: add link participant for linkage transitions

### PR DESCRIPTION
**Pull request by zyga**
The export manager needs to perform bookkeeping / updates on the new
current symbolic link that it is introducing. Instead of adding a new
task that is conceptually very similar to link-snap, and risking
breaking the complex sequence and dependencies between tasks, add a way
to hook into link-snap itself.

The link participant is a callback which can be used to obtain
notifications about link-snap, unlink-snap as well as their undo
executions.

Instead of carrying details as to how linkage changed, link participant
offers the system state and a snap name, allowing interested parties to
register a participant that can recover as little or as much information
as is required.

Since link-snap is so critical to the execution logic, link snap
participant failures are non-binding. The notification is only
dispatched after link-snap has effectively succeeded, save for making
the task as complete. Similar approach is used for unlink-snap as well
as the undo paths.

In effect a participant can be used to follow linkage changes and update
ancillary information, but cannot impact the overall change.

Signed-off-by: Zygmunt Krynicki <me@zygoon.pl>

### CODE REVIEW

**Comment by pedronis** [comment]
@zyga in isolation it's strange to me that is not called by unlink-current-snap and its undo ?

As we know I'm also not sure about this not erroring but that's less of an immediate concern to landing this

**Comment by zyga** [comment]
@pedronis unlink-current-snap is an omission and I agree it should be changed. The error question is a bigger discussion and I think it'd be fair to rename this to LinkObserver, as that coveys the error behavior more appropriately.

File: overlord/snapstate/handlers.go [hunk 1]
```diff
@@ -923,6 +926,8 @@ func (m *SnapManager) undoUnlinkCurrentSnap(t *state.Task, _ *tomb.Tomb) error {
 	// so that we switch executing its snapd
 	m.maybeRestart(t, oldInfo, reboot, deviceCtx)
 
+	// Notify link snap participants about link changes.
+	notifyLinkParticipants(t, snapsup.InstanceName())
```
Comment (pedronis) [inline overlord/snapstate/handlers.go:930]: for consitency I would do this before the maybeRestart

File: overlord/snapstate/handlers.go [hunk 2]
```diff
@@ -1236,6 +1282,7 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) (err error) {
 		if unlinkErr != nil {
 			t.Errorf("cannot cleanup failed attempt at making snap %q available to the system: %v", snapsup.InstanceName(), unlinkErr)
 		}
+		notifyLinkParticipants(t, snapsup.InstanceName())
```
Comment (pedronis) [inline overlord/snapstate/handlers.go:1285]: we have a preexisting bug where it seems nowadays we can get here after we have Set the snapstate, this means we can end up calling the link participant with Active = true, but we just unlinked the snap again...

bit of a mess that we need to untangle somehow

File: overlord/snapstate/handlers.go (hunk 1 above)
Comment (zyga) [reply overlord/snapstate/handlers.go:930]: Done

File: overlord/snapstate/handlers.go (hunk 2 above)
Comment (zyga) [reply overlord/snapstate/handlers.go:1285]: I understand this is solved by your branch which is now merged here.

**Review by pedronis — APPROVED** [review]
thanks, on its own terms this looks alright

**Review by mvo5 — APPROVED** [review]
Looks good, thanks for your work on this!

**Pull request merged**